src/graph.py

```diff
@@ -43,46 +57,111 @@ def edges(self):
-            self._dict[str(n1)].remove(str(n2))
-        except KeyError:
-            raise ValueError('No such edge exists')
+            self._dict[n1].remove(n2)
+        except KeyError as e:
+            e.args = ('No such edge exists',)
+            raise
 
     def has_node(self, n):
-        """Check if n is a node of graph"""
-        if str(n) in self._dict.keys():
+        """Check if n is a node of graph."""
+        if n in self._dict.keys():
             return True
         else:
             return False
 
     def neighbors(self, n):
-        """Return a list of nodes that have edge connect to n"""
+        """Return a list of nodes that have edge connect to n."""
         try:
-            self._dict[str(n)]
-            neighbors = []
-            for pair in self._dict.items():
-                if str(n) in pair[1]:
-                    neighbors.append(pair[0])
-            return neighbors
-        except KeyError:
-            raise ValueError('Node not in the graph')
+            return self._dict[n]
+        except KeyError as e:
+            e.agrs = ('Node not in the graph',)
+            raise
 
     def adjacent(self, n1, n2):
-        """Check if 2 node has connection"""
+        """Check if 2 nodes has connection."""
         try:
-            return str(n2) in self._dict[str(n1)] or\
-             str(n1) in self._dict[str(n2)]
-        except KeyError:
-            raise ValueError('Node not in the graph')
+            return n2 in self._dict[n1] or\
+             n1 in self._dict[n2]
+        except KeyError as e:
+            e.agrs = ('Node not in the graph',)
+            raise
+
+    def depth_first_traversal(self, start):
+        """
+        Perform a full depth-traversal of the graph beggining at start.
+        Return full visited path when traversal is complete.
+        Raise a ValueError, if the graph is empty.
+        """
+        if self._dict == {}:
+            raise ValueError("Can't traverse an empty graph.")
+        path_list = [start]
+        visited_list = [start]
+        current_node = start
+        while current_node:
+            for n in self.neighbors(current_node):
+                if n not in path_list:
+                    path_list.append(n)
+                    visited_list.append(n)
+                    current_node = n
+                    break
+            else:
+                try:
+                    visited_list.pop()
+                    current_node = visited_list[-1]
+                except IndexError:
+                    break
+        return path_list
+
+    def breadth_first_traversal(self, start):
+        """
+        Perform a full breadth-traversal of the graph beggining at start.
+        Return full visited path when traversal is complete.
+        Raise a ValueError, if the graph is empty.
+        """
+        if self._dict == {}:
+            raise ValueError("Can't traverse an empty graph.")
+        path_list = [start]
+        pending_list = []
```

Comment:
This is the right idea -- keeping a data structure which is your "nodes to visit", like a to-do list. The difference between a BFS and a DFS is only whether your to-do list is first-in-first-out or first-in-last-out.